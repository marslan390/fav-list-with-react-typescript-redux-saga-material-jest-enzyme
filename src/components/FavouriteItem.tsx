import React from "react";
import Rating from "@material-ui/lab/Rating";
import { makeStyles } from "@material-ui/core/styles";
import Card from "@material-ui/core/Card";
import CardActionArea from "@material-ui/core/CardActionArea";
import CardContent from "@material-ui/core/CardContent";
import Typography from "@material-ui/core/Typography";
import { FavItem } from "../store/fav-list/types";

const useStyles = makeStyles({
  card: {
    borderRadius: 0
  }
});

interface FavouriteItemProps {
  item: FavItem;
  updateRating: (favItem: FavItem, rating: number) => void;
}

const FavouriteItem: React.FC<FavouriteItemProps> = ({
  item,
  updateRating
}) => {
  const classes = useStyles();

  const onChangeRating = (event: any, rating: number | null) => {
    if (rating) {
      const favItem = { ...item };
      updateRating(favItem, rating);
    }
  };

  return item.rating > 0 && item.rating <= 5 ? (
    <Card className={classes.card}>
      <CardActionArea>
        <CardContent>
          <Typography
            data-testid="name"
            gutterBottom
            variant="h5"
            component="h2"
          >
            {item.name}
          </Typography>
          <Typography
            data-testid="author"
            gutterBottom
            variant="body2"
            color="textSecondary"
            component="p"
          >
            {item.author}
          </Typography>
          <Typography component="p">
            <Rating
              name={`rating${item.id}`}
              size="large"
              value={item.rating}
              precision={0.5}
              onChange={onChangeRating}
              max={5}
            />
          </Typography>
        </CardContent>
      </CardActionArea>
    </Card>
  ) : (
    <span data-testid="invalidRating">Invalid Rating</span>
  );
};
export default FavouriteItem;
