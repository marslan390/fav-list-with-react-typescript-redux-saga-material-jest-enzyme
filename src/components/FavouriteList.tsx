import React from "react";
import { makeStyles } from "@material-ui/core/styles";
import FavouriteItem from "./FavouriteItem";
import { FavItem } from "../store/fav-list/types";
import Pagination from "@material-ui/lab/Pagination";
import Card from "@material-ui/core/Card";
import usePagination from "../hooks/pagination";

const useStyles = makeStyles({
  card: {
    borderRadius: 0,
    borderTop: "1px solid #f0f0f0",
    padding: "5px 0"
  }
});

interface FavouriteListProps {
  items: FavItem[];
  isRating: boolean;
  toggleRating: () => void;
  updateRating: (favItem: FavItem, rating: number) => void;
}

const FavouriteList: React.FC<FavouriteListProps> = ({
  items,
  updateRating
}) => {
  const classes = useStyles();
  const [pagingProps, pageItems] = usePagination(items, 5);
  return items.length > 0 ? (
    <>
      {pageItems.map((favItem: FavItem) => (
        <FavouriteItem
          key={favItem.id}
          item={favItem}
          updateRating={updateRating}
        />
      ))}
      <Card className={classes.card}>
        <Pagination {...pagingProps}/>
      </Card>
    </>
  ) : (
    <span data-testid="notFound">No item found</span>
  );
};

export default FavouriteList;
