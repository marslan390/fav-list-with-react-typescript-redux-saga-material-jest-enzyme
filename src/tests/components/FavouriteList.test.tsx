import React from "react";
import {mount, shallow} from "enzyme";
import FavouriteList from "../../components/FavouriteList";
import FavouriteItem from "../../components/FavouriteItem";

describe("<FavouriteList/>", () => {
  let props: any;

  beforeEach(() => {
    props = {
      items: [
        {
          id: 1,
          name: "The Hunger Games",
          author: "Suzanne Collins",
          rating: 4.33
        },
        {
          id: 2,
          name: "Harry Potter and the Order of the Phoenix",
          author: "J.K. Rowling",
          rating: 4.49
        },
        {
          id: 3,
          name: "To Kill a Mockingbird",
          author: "Harper Lee",
          rating: 4.27
        },
        {
          id: 4,
          name: "Pride and Prejudice",
          author: "Jane Austen",
          rating: 4.26
        },
        {
          id: 5,
          name: "Twilight",
          author: "Stephenie Meyer",
          rating: 4.33
        }
      ],
      updateRating: jest.fn( () => {})
    };
  });

  test("renders correctly", () => {
    const wrapper = shallow(<FavouriteList {...props} />);
    expect(wrapper).toMatchSnapshot();
  });

  test("renders fav item List", () => {
    const wrapper = mount(<FavouriteList {...props} />);
    expect(wrapper.find(FavouriteItem)).toHaveLength(5);
  });

  test("renders empty fav list", () => {
    props = { ...props, items: [] };
    const wrapper = mount(<FavouriteList {...props} />);
    expect(
      wrapper.find({ "data-testid": "notFound" }).hostNodes()
    ).toHaveLength(1);
  });
});
