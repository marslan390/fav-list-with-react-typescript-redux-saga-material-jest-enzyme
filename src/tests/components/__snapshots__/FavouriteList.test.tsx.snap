// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<FavouriteList/> renders correctly 1`] = `ShallowWrapper {}`;
